### ISSUE TITLE
pandoc loses local formatting

### ISSUE THREAD

**Issue by nehrwert** [jgm/pandoc#2623]
Hi, 

I encountered a strange behavior in the conversion from odt to html/md (in the sense that I don't understand the reasoning): If a whole line is locally formatted (i.e. bold or italic) pandoc "forgets" the formatting, if it is formatted with a character style it is formatted in the produced output. 

Should pandoc behave this way? Can you disable this to preserve local formatting?
I'm not sure if this is the correct space for this question, if not, I'm sorry!

**Comment by jgm** [comment]
@MarLinn wrote the ODT reader and might be able to help.

**Comment by MarLinn** [comment]
That does sound like a bug.

Do I understand correctly that by "locally formatted" you mean you are changing the paragraph style?

If so, then the problem might be in converting paragraph styles to character styles inside the odt reader. If I remember correctly (it's been a while) Pandoc has no support for paragraph-level formatting like this internally, so paragraph styles must be converted to character styles. That also means that if this is a bug the end result after fixing it will be the same as if you had applied a character style, even if the output format supports paragraph level formatting. Still, better to preserve the style in the wrong way than to ignore it completely.

Do you have a minimal example I could use as a test case?

**Comment by nehrwert** [comment]
Thanks for your answers:

"Do I understand correctly that by "locally formatted" you mean you are changing the paragraph style?"

No, with locally formatted I mean, that I pushed the bold or italic button. (as opposed to using f.e. the emphasis character style). I have created a small example, which you can find here:
https://depot.free.de/files/h8NjHWyCrQ8 

In it there are two lines italic and one bold, the last italic one is formatted with the emphasis character style.

**Comment by mb21** [comment]
The link no longer contains the file, please upload it directly to this GitHub issue. Until then, I'm closing this as it's not much use.

**Comment by caaaaaa** [comment]
[pandoctest.zip](https://github.com/jgm/pandoc/files/3439557/pandoctest.zip)

I have encountered the same bug, there are some test cases in an .odt document, (zipped for github) produced by libreoffice Version: 6.1.6.3.

The text cursive in an paragraph becomes *cursive* in the resulting .md file, while the whole cursive paragraph (double click paragraph to select, press Italic button) does not. Bold text gets formatted as expected.

Output by pandoc 2.7.3 (2.1.2 is the same)

This is a normal Paragraph. This is a normal Paragraph. This is a normal
Paragraph. This is a normal Paragraph.

This is a normal Paragraph. *But something is cursive.* This is a normal
Paragraph. **And something bold.** This is a normal Paragraph.

This is a normal Paragraph.

This is a cursive paragraph.

This is a normal Paragraph. This is a normal Paragraph.

**This is a bold paragraph.**

This is a normal Paragraph. This is a normal Paragraph.

**Comment by caaaaaa** [comment]
I have found another, maybe related issue with cursive text in odt files. 
After cursive words it should continue normal, but inserts ** when converted to markdown, making the whole rest of the document appear bold. 
I have copied this example out of a document i received for publishing, when i made something similar i could not reproduce this, so it seems a property of the specific odt formatting. 
When i safe this example in libreoffice as docx and convert it with pandoc, everything is fine
[test.zip](https://github.com/jgm/pandoc/files/3685945/test.zip)

**Comment by UnixRonin** [comment]
Another thing that seems to **consistently** reproduce this problem is if you select several paragraphs at once in an ODT document, italicise or boldface them, and then convert the result to HTML.  Formatting that encloses *multiple paragraphs* seems to just get forgotten.

I'm not **COMPLETELY** certain yet, but it **SO FAR APPEARS** that a workaround for this is to save a copy of the ODT as DOCX, **then** use pandoc to convert from that DOCX to HTML/EPUB.  This appears to support the hypothesis that the problem lies in the ODT reader.


**Separately,** I have also noticed that small-caps formatting just ...  vanishes.  No matter where it is.

**Comment by jgm** [comment]
See #10575 for further discussion.

> what is going on here is that a (temporary) *paragraph* style is being used that sets text properties to italic, and then a separate text style is used to reset things to normal. The ODT reader isn't picking up on this. It does okay with the more normal situation in which a *text* style sets the text to italics.

**Comment by mxamber** [comment]
Just ran into the same issue.

What happens is that in the content.xml file inside the .odt archive, each paragraph and each differently formatted length of text is assigned its own style. The first paragraph has paragraph style P1, the next paragraph style P2, ... Pn, even if all of them are identical in formatting. And likewise, marking a single word italics or bold gives that length of text (represented internally with a `<text:span>` XML element its own text style: T1, T2, ... Tn.

This is further complicated by ODT's versioning / track changes feature, which, every time you go back to fix a typo, encloses that typo in another text:span element with an rsid value to track changes. Which, slight tangent, leads to extremely irritating behaviour every time you paste LO rich text into any rich text interface elsewhere and words get broken in half.

As established upthread, Pandoc only considers the text styles. So if an entire paragraph is selected and italicised in LO, its *paragraph* style is changed (instead of wrapping the entire paragraph in a text:span element with an italicised *text* style) and Pandoc doesn't consider paragraph styles.

**However**, it's not just ODT, so I'm not sure if the issue is with the ODT reader. Because, in my quest to find a workaround, I first tried saving from LO as HTML (terrible decision; it produces garbage HTML), and then exporting to XHTML. The resulting file preserved the paragraph styles, in the form of framkly indecipherable "don't use any semantic tags and instead slap its own class on each paragraph and style those", and running that through Pandoc to convert to i.e. Markdown *also* resulted in italicised paragraphs being ignored.

So either the issue lies deeper, or the HTML parser and ODT parser both independently contain the same issue.

**Comment by UnixRonin** [comment]
Yeah, LO's internal handling of formatting is ghastly.  Every time you backspace or fix a typo, it inserts a new span.  One of my first steps **towards** cleaning it up was to write a Perl tool that unpacked an LO-generated EPUB and stripped *only* all of the **redundant** span tags out of the XHTML.  That cut the file size of many XHTML sections by literally 50%.  It turned out to be less work to manually strip out **ALL** of LO's formatting from an EPUB, hand-write new CSS, and go back through the entire book reformatting everything by hand, than to try to fix LO's (or libepubgen's) horribly broken XHTML.

**Comment by Alayan-stk-2** [comment]
I encountered this issue too, italic formatting spanning multiple paragraphs (and present through the italic button, not text style) dropped by the ODT reader and missing from the AST intermediate file.

**Comment by jgm** [comment]
Here's a small test case that illustrates the issue.
[2623.odt](https://github.com/user-attachments/files/32587037/2623.odt)
Result from pandoc:
```
<p>Italics</p>
<p>not <em>italics</em></p>
```
The first paragraph should be entirely italics, but this isn't picked up.

**Comment by jgm** [comment]
Relevant XML:
```xml
  <office:automatic-styles>
    <style:style style:name="P1" style:family="paragraph" style:parent-style-name="Standard">
      <style:text-properties fo:font-style="italic" officeooo:rsid="001a850e" officeooo:paragraph-rsid="001a850e" style:font-style-asian="italic" style:font-style-complex="italic"/>
    </style:style>
    <style:style style:name="P2" style:family="paragraph" style:parent-style-name="Standard">
      <style:text-properties fo:font-style="normal" officeooo:rsid="001a850e" officeooo:paragraph-rsid="001a850e" style:font-style-asian="normal" style:font-style-complex="normal"/>
    </style:style>
    <style:style style:name="T1" style:family="text">
      <style:text-properties fo:font-style="italic" style:font-style-asian="italic" style:font-style-complex="italic"/>
    </style:style>
  </office:automatic-styles>
  <office:body>
    <office:text>
      <text:sequence-decls>
        <text:sequence-decl text:display-outline-level="0" text:name="Illustration"/>
        <text:sequence-decl text:display-outline-level="0" text:name="Table"/>
        <text:sequence-decl text:display-outline-level="0" text:name="Text"/>
        <text:sequence-decl text:display-outline-level="0" text:name="Drawing"/>
        <text:sequence-decl text:display-outline-level="0" text:name="Figure"/>
      </text:sequence-decls>
      <text:p text:style-name="P1">Italics
      </text:p>
      <text:p text:style-name="P2"/>
      <text:p text:style-name="P2">not
      <text:span text:style-name="T1">italics
      </text:span>
      </text:p>
      <text:p text:style-name="P2"/>
    </office:text>
  </office:body>
```
So, I am guessing that what is happening here is quite simple. ODT can put inline-level formatting, e.g. font-style, on a paragraph style, but the ODT reader (I'm conjecturing) is only reading the block-level formatting from paragraph styles. If so, this certainly seems fixable.

**Comment by jgm** [comment]
Surprisingly easy to fix!

**Comment by Alayan-stk-2** [comment]
Ah! I didn't expect my comment on a 10-year old report to trigger a fix so quickly!

Good job and thank you!
